packages/client/src/util/toDataValueQueries.ts

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2024 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type { QueryDataTypeDefinition } from "@osdk/api";
+import { type DataValue, Ontologies } from "@osdk/internal.foundry";
+import type { MinimalClient } from "../MinimalClientContext.js";
+import { isAttachmentUpload } from "../object/AttachmentUpload.js";
+import { getWireObjectSet, isObjectSet } from "../objectSet/createObjectSet.js";
+import { isOsdkBaseObject } from "./isOntologyObjectV2.js";
+import { isWireObjectSet } from "./WireObjectSet.js";
+
+/**
+ * Marshall user-facing data into the wire DataValue type
+ *
+ * @see DataValue for the expected payloads
+ */
+export async function toDataValueQueries<T extends string>(
+  value: unknown,
+  client: MinimalClient,
+  desiredType: QueryDataTypeDefinition<T>,
+): Promise<DataValue> {
+  if (value == null) {
+    return value;
+  }
+
+  if (Array.isArray(value) && desiredType.multiplicity) {
+    const promiseArray = Array.from(
+      value,
+      async (innerValue) =>
+        await toDataValueQueries(innerValue, client, desiredType),
+    );
+    return Promise.all(promiseArray);
+  }
+
+  switch (desiredType.type) {
+    case "attachment": {
```

Comment:
Now that we have the metadata of the query to help us convert to the right wire type, I'm not sure if we should throw if we don't get the right types. E.g. if the desiredType is an object, but what we got is not an `OsdkBaseObject`, should we throw here, or just pass through the value to the wire and rely on the backend to tell us something went wrong? I'm leaning on the side of us throwing here, but not sure

Reply:
Behaviorally I think they will both result in an error thrown, it is just a question of how useful the error is. This might be a situation where in production we just pass through but in development we provide a lot of details. Once you get to production this type of logical error shouldnt exist.